File: balanced_delimiter/solutions/balanced_delimiter.js
var closersToOpeners = {
  ']': '[',
  ')': '(',
  '}': '{'
};

function isCloser(character) {
  return closersToOpeners.hasOwnProperty(character);
}

function isOpener(character) {
  for (var key in closersToOpeners) {
    if (closersToOpeners.hasOwnProperty(key) && closersToOpeners[key] === character) {
      return true;
    }
  }

  return false;
}

function isBalanced(delimiterString) {
  var openerStack = [];

  for (var i = 0; i < delimiterString.length; i++) {
    var currentChar = delimiterString[i];

    if (isOpener(currentChar)) {
      openerStack.push(currentChar);
    } else if (isCloser(currentChar)) {
      var expectedOpener = openerStack.pop();
      if (closersToOpeners[currentChar] !== expectedOpener) {
        return false;
      }
    } else {
      throw "Non-delimiter character in input";
    }
  }

  return openerStack.length === 0;
}

console.log(isBalanced("{}()[]"));
console.log(isBalanced("([)]"));
console.log(isBalanced("([omg)]")); // exception

File: balanced_delimiter/solutions/balanced_delimiter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

class balanced_delimiter
{
    static readonly Dictionary<char, char> delimiterLookup = new Dictionary<char, char>()
     {
         { '(', ')'},
         { '[', ']'},
         { '{', '}'}
     };

    static void Main(string[] args)
    {
        const string delimiters = "([])";
        Console.WriteLine(IsBalanced(delimiters));
    }

    public static bool IsBalanced(string delimiters)
    {
        var delimiterCharArray = delimiters.ToCharArray();
        var delimiterStack = new Stack<string>();

        foreach (var delimiter in delimiterCharArray)
        {
            if (IsOpener(delimiter))
            {
                delimiterStack.Push(delimiter.ToString());
            }
            else if (!IsOpener(delimiter))
            {
                if (!IsMatchingCloser(delimiter, delimiterStack.Pop())) return false;
            }
            else
            {
                throw new InvalidDataException("Wrong input!");
            }
        }

        return !delimiterStack.Any();
    }

    private static bool IsMatchingCloser(char possibleMathcingCloser, string opener)
    {
        return delimiterLookup.FirstOrDefault(d => d.Key == Convert.ToChar(opener)).Value == possibleMathcingCloser;
    }

    private static bool IsOpener(char delimiter)
    {
        return delimiterLookup.ContainsKey(delimiter);
    }
}

